### ISSUE TITLE
Deserializer doesn't close level with single-segment group

### ISSUE THREAD

**Issue by GVG** [indice-co/EDI.Net#149]
Hello! I have an issue with EDIFACT deserializer when it doesn't close level if group consists of single segment.
Fragment of EDI document:
```
...
AJT+ZZZ+35'
TAX+4'
PAI+::1:1:1'
RFF+AAA:X'
MOA+ZZZ:9:CAD'
UNS+D'
...
```
According to partner's documentation AJT is segment group 15 and the TAX+PAI+RFF+MOA are next group 25. Models:
```csharp
    [EdiMessage]
    public class Message
    {
        ...
        public SG15 Aaa { get; set; }

        public SG25 Bbb { get; set; }
        ...
    }

    [EdiSegmentGroup("AJT", SequenceEnd = "UNS")]
    public class SG15
    {
        [EdiValue("X(3)", Path = "AJT/0", Mandatory = true)]
        public string Aaa { get; set; }

        [EdiValue("X(2)", Path = "AJT/1")]
        public string Bbb { get; set; }
    }

    [EdiSegmentGroup("TAX", "PAI", "RFF", "MOA", SequenceEnd = "UNS")]
    public class SG25
    {
        [EdiValue("X(3)", Path = "TAX/0", Mandatory = true)]
        public string Aaa { get; set; }

        public List<SG26> Bbb { get; set; }
    }
```

I tried debugging and here is what I've found out:
1. Deserializer creates new level for SG15
2. Reading AJT as member of SG15.
3. Reading TAX, PAI, RFF, MOA as member of SG15 (does nothing as there are no such segments in SG15).
4. Reaches UNS and closes level.

There are 3 conditions in `EdiSerializer.TryCreateContainer(...)` which close level and normally it will close level on step 3 because processed level doesn't contain TAX segment according to SG15 attribute (condition 3). But this doesn't happen in this case since SG15 contains only 1 segment group member: `level.GroupMembers.Length > 1 && !level.GroupContains(readerSegment as string)`
Instead it keeps reading following segments on level SG15 until it reaches sequence end segment (condition 2).
Adding any random segments to SG15 attribute "solves" this issue. `[EdiSegmentGroup("AJT", "XXX", SequenceEnd = "UNS")]`

So basically a segment group with only one member makes deserializer go crazy till it reaches sequence end segment if defined or end of the document.
Is this a bug?

**Comment by cleftheris** [comment]
Hi @GVG 
This sounds like a bug. Can you please send me the model classes and a sample transmission so I can add a test for this. If you feel like it you can do so yourself and make a pull request so I can investigate further. 

On an other note (probably not related) the sequenceend property on the SegmentGroupAttribute is not designed to work with multiple segment names. Either you define all segment names with params or use sequencestart and sequenceend (escape)

**Comment by GVG** [comment]
@cleftheris should I do pull request with samples in `test/indice.Edi.Tests` for test?
Not sure how to fix this. What was the idea behind condition `level.GroupMembers.Length > 1`? In my case there is only one segment defined in the attribute, but deserializer understands it as defined sequence start with unknown total number of segments and defined sequence end. I'm afraid removing this condition can break something.
Regarding sequence end property with explicitly defined segments. Personally I don't see any issue here since these are 2 parallel level-closing conditions. Actually I added it later.
I think this can be fixed by changing condition 2 and 3 like:
1. Same. No idea when condition 1 is true (can you explain?)
2. `else if (!sequenceEnd.HasValue && !level.GroupContains(readerSegment as string))`
3. `else if (sequenceEnd.HasValue && sequenceEnd.Value.Segment.Equals(readerSegment))`

Basically in this variant defining sequence end means that you didn't specify whole list explicitly. Must ensure this won't become a breaking change.

**Comment by cleftheris** [comment]
No need to fix something in order to become a contributor. Just the samples in the test project would be fine. 

> Basically in this variant defining sequence end means that you didn't specify whole list explicitly. Must ensure this won't become a breaking change.

Exactly. SequenceEnd works with the start segment only. In 99.9% of the cases you are better of explicitly defining the list of segments and not using it.

**Comment by GVG** [comment]
Created pull request with samples and comments. IMO this behavior should be covered in the documentation with examples.

**Comment by cleftheris** [comment]
@GVG I agree with you on that. I must find some time to create a wiki with examples but I have not found the time to do so. I am always open to contributions.

Thanks for your time. I will be looking into the pull request

**Comment by cleftheris** [comment]
Ok now I see the issue more clearly. So the SG15 class should not be a segmentgroup. You only need a AJT class there decorated with the EdiSegment and EdiPath attributes. As far as the library is concerned segment groups are defined when there are more than one segments involved.

**Comment by cleftheris** [comment]
@GVG 
Should I be throwing an exception when there is a segmentgroup attribute with a single segment name there or support the scenario even though it is equivalent to a segment attribute. I think the first option is better but may break some people. The second option feels redundant. What are your thoughts?

**Comment by GVG** [comment]
In my opinion ideally it should support single-segment groups or even scan whole structure and create group members list automatically, but this change can cause issues in older code. Probably for the sake of backward compatibility this can be "fixed" by throwing an exception. The question is should sequence end parameter become mandatory now if you specify only first segment? If yes then probably throw something like "For single-segment group use EdiSegment attribute or specify sequence end segment." and mention this in the doc. Also mention that you ideally should list all segments manually and sequence end segment is next segment **after** the group (this is unclear for new users).

PS is there some technical reason we have to list group members manually?

**Comment by cleftheris** [comment]
Thanks for the feedback @GVG .

Regarding your question the reason you need to explicitly define all segments has to do with the stractureless nature of the EDI format. As the library matured over time there where numerous corner cases where the groups where encapsulating other groups containing the same segment names in different levels with or without conditions. Check some of the closed issues https://github.com/indice-co/EDI.Net/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+segmentgroup  especially the #91 which was a huge advancement forward

**Comment by GVG** [comment]
@cleftheris Thank you!
